client/src/sync-filesystem.js

```diff
@@ -6,39 +6,216 @@
-      if(path) {
-        pathToSync = resolvePath(pathToSync, path);
-      } else {
-        pathToSync = null;
-      }
+  var root = '/';
+  // Record changes during a downstream sync
+  // Is a parallel array with sourceListCache
+  var recordedPaths = [];
+  var sourceListCache = [];
+  var changesDuringDownstream = [];
+
+  // Expose the root used to sync for the filesystem
+  // Defaults to '/'
+  Object.defineProperties(self, {
+    'root': {
+      get: function() { return root; }
+    },
+    'changesDuringDownstream': {
+      get: function() { return changesDuringDownstream; }
     }
   });
 
-  // Record modifications to the filesystem during a sync
-  Object.defineProperty(fs, 'modifiedPath', {
-    get: function() { return modifiedPath; },
-    set: function(path) {
-      if(path) {
-        modifiedPath = resolvePath(modifiedPath, path);
-      } else {
-        modifiedPath = null;
+  self.record = function(path, sourceList) {
+    recordedPaths.push(path);
+    sourceListCache.push(sourceList);
+  };
+
+  self.stopRecording = function(path) {
+    var indexInRecordedPaths = recordedPaths.indexOf(path);
+    recordedPaths.splice(indexInRecordedPaths, 1);
+    changesDuringDownstream.splice(changesDuringDownstream.indexOf(path), 1);
+    return sourceListCache.splice(indexInRecordedPaths, 1)[0];
+  };
+
+  // Get the paths queued up to sync
+  self.getPathsToSync = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
       }
+
+      callback(null, pathsToSync && pathsToSync.toSync);
+    });
+  };
+
+  // Add paths to the sync queue where paths is an array
+  self.appendPathsToSync = function(paths, callback) {
+    if(!paths || !paths.length) {
+      return callback();
     }
-  });
+
+    paths = paths.filter(function(path) {
+      return path.indexOf(root) === 0;
+    });
+
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      var toSync = pathsToSync.toSync;
+
+      // Ignore redundancies
+      paths.forEach(function(path) {
+        if(toSync.indexOf(path) === -1) {
+          pathsToSync.toSync.push(path);
+        }
+      });
+
+      fsUtils.setPathsToSync(fs, root, pathsToSync, callback);
+    });
+  };
+
+  // Get the path that was modified during a sync
+  self.getModifiedPath = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      callback(null, pathsToSync && pathsToSync.modified);
+    });
+  };
+
+  // Indicate that the path on top of the queue has
+  // begun syncing
+  self.setSyncing = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      if(!pathsToSync || !pathsToSync.toSync || !pathsToSync.toSync[0]) {
+        return callback();
```

Comment:
Should we warn here?  It might indicate a bug in our logic if there actually aren't any paths in the queue.